cluster/engine.go

```diff
@@ -904,22 +927,41 @@ func (e *Engine) String() string {
 
 func (e *Engine) handler(ev *dockerclient.Event, _ chan error, args ...interface{}) {
 	// Something changed - refresh our internal state.
-	switch ev.Status {
-	case "pull", "untag", "delete", "commit":
-		// These events refer to images so there's no need to update
-		// containers.
-		e.RefreshImages()
-	case "die", "kill", "oom", "pause", "start", "stop", "unpause", "rename":
-		// If the container state changes, we have to do an inspect in
-		// order to update container.Info and get the new NetworkSettings.
-		e.refreshContainer(ev.ID, true)
-		e.RefreshVolumes()
+
+	switch ev.Type {
+	case "network":
 		e.RefreshNetworks()
-	default:
-		// Otherwise, do a "soft" refresh of the container.
-		e.refreshContainer(ev.ID, false)
+	case "volume":
 		e.RefreshVolumes()
-		e.RefreshNetworks()
+	case "image":
+		e.RefreshImages()
+	case "container":
+		switch ev.Action {
+		case "die", "kill", "oom", "pause", "start", "stop", "unpause", "rename":
```

Comment:
When you remove a container, docker will either stop or kill it before and it will trigger a `network disconnect` event.
So I don't think a `RefreshNetworks` is needed.